### PR TITLE
BUGFIX: TSDB: panic in query during truncation with OOO head

### PR DESCRIPTION
**Pull request by krajorama**
Added regression test for #14822. Doesn't cause segfault before #14354

The segfault was due to a race condition between query start and compaction.
When compaction starts, in-order queries may overlap with the TSDB head, but also might fall into the truncated time of the head. In such case, the head querier `headQuerier` is nil in db.go here
https://github.com/prometheus/prometheus/blob/1e01e97151c9db59178dfbabd5e60f4b3916cb83/tsdb/db.go#L2062
and
https://github.com/prometheus/prometheus/blob/1e01e97151c9db59178dfbabd5e60f4b3916cb83/tsdb/db.go#L2137

That pointer is not used for selecting samples, but is referenced in `Close()` which causes the segfault.

The fix essentially restores the original function where we did not rely on the headQuerier in creating the OOO head querier: https://github.com/prometheus/prometheus/blob/9849418fac06ae0c0d3999891940939d8e0f8307/tsdb/db.go#L2157

### CODE REVIEW

File: tsdb/ooo_head_read.go [hunk 1]
```diff
@@ -513,7 +513,7 @@ type HeadAndOOOQuerier struct {
 	head       *Head
 	index      IndexReader
 	chunkr     ChunkReader
-	querier    storage.Querier
+	querier    storage.Querier // This might be nil if head was truncated.
```
Comment (bboreham) [inline tsdb/ooo_head_read.go:516]: Generally I advise to state what the thing means, not when you expect it to apply.
So "If nil, do not read from in-order head"?

Comment (krajorama) [reply tsdb/ooo_head_read.go]: clarified

**Comment by krajorama** [comment]
cc PTAL @colega

**Review by jesusvazquez — APPROVED** [review]
LGTM! I can see how you are preventing new panics and how data is still coming from the blocks in the test. Nice work!

**Review by bboreham — APPROVED** [review]

**Pull request merged**